Updates for Windows Server version 1709 with K8s v1.8

JMesser81 opened the pull request:
This updated getting-started guide targets Windows Server version 1709 (GA October 2017) using Kubernetes v1.8.2 with several additional PRs required for building kubelet and kube-proxy to work on Windows. The guide references different supported network topologies and available CNI plugins. 

Links to Microsoft/SDN repo reference instructions needed for building windows K8s binaries locally as well as instructions for running scripts to join Windows node to an existing K8s cluster.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/website/6180)
<!-- Reviewable:end -->

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
-   1. Linux/MacOS: `KUBE_BUILD_PLATFORMS=windows/amd64 make WHAT=cmd/kube-proxy`
-   2. Windows: `go build cmd/kube-proxy/proxy.go`
+2. Configure Linux Master node using steps [here](https://github.com/Microsoft/SDN/blob/master/Kubernetes/HOWTO-on-prem.md)
+3. [Optional] CNI network plugin installed.
```
This is only partially true, right? By default (i.e. using `cni=kubenet`), you need at least the `host-local`, `loopback`, and `bridge` CNI plugins. And not every third-party CNI will work with Windows nodes. Maybe this is implied and known, but still thought I'd mention it.

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+4. Create HNS Network
+5. Ensure correct CNI network config
+5. Start kubelet.exe using this script [start-kubelet.ps1](https://github.com/Microsoft/SDN/blob/master/Kubernetes/windows/start-kubelet.ps1)
+6. Start kube-proxy using this script[start-kubeproxy.ps1](https://github.com/Microsoft/SDN/blob/master/Kubernetes/windows/start-kubeproxy.ps1)
```
nit: space after "script"

gkudra-msft replied:
If you're starting using the script, you don't need steps 4 & 5. I like that it basically outlines what the script will do so it's not a black box, but maybe it should just be inline with the kubelet step.

bsteciuk replied:
addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+**Windows CNI Config Example**
+Today, Windows CNI plugin is based on wincni.exe code with the following example, configuration file.
 
+Note: this file assumes that a user previous created 'l2bridge' host networks on each Windows node using <Verb>-HNSNetwork cmdlets as shown in the start-kublet.ps1 and start-kubeproxy.ps1 scripts linked above
```
nit: minor grammar issues, also `<Verb>` is getting swallowed by the viewer (as HTML, I think?) so it just says "using -HNSNetwork." Escape it or wrap it in back-ticks. I'd wrap the script names in back-ticks too, for aesthetics :)

rjmorse replied:
ensure `start-kublet.ps1` is changed to `start-kubelet.ps1` , since the file was originally `start-kublet.ps1`

bsteciuk replied:
Fixed these in https://github.com/JMesser81/kubernetes.github.io/pull/1

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+                                         }
+                           }
+                       ]
+}
```
You could run the JSON through [this linter](https://jsonlint.com/) so that the spacing is prettier. :)

gkudra-msft replied:
Ah, I see the issue. There's a newline missing right before the code block.

bsteciuk replied:
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
-2. Run *kube-proxy* executable using the below command
-`.\proxy.exe --v=3 --proxy-mode=userspace --hostname-override=<ip address/hostname of the windows node> --master=<api server location> --bind-address=<ip address of the windows node>`
+## Starting the Linux-based Control Plane
+Use your preferred method to setup and start Kubernetes cluster on Linux or follow the directions given in this [link](https://github.com/Microsoft/SDN/blob/master/Kubernetes/HOWTO-on-prem.md). Please note that Cluster CIDR might need to be updated.
```
...since it's hard-coded to `11.0.0.0/16` in the scripts. Should we make that parameterized?

Also, though it's in the linked guide, reiterate the necessity of copying the certificate configuration file from the master.

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+Use your preferred method to setup and start Kubernetes cluster on Linux or follow the directions given in this [link](https://github.com/Microsoft/SDN/blob/master/Kubernetes/HOWTO-on-prem.md). Please note that Cluster CIDR might need to be updated.
 
 ## Scheduling Pods on Windows
 Because your cluster has both Linux and Windows nodes, you must explicitly set the nodeSelector constraint to be able to schedule Pods to Windows nodes. You must set nodeSelector with the label beta.kubernetes.io/os to the value windows; see the following example:
```
nit: backticks on `nodeSelector` and the label, pods caps

Also, you can write 

    ```yaml
    [the YAML file]
    ```

to get syntax highlighting! Likewise, use `json` for the CNI config file. 😄

gkudra-msft replied:
Update: looks like they actually auto-detected the YAML it in the deployment, but the CNI config is broken for some reason. Either way, it'll make the GitHub preview prettier.

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
 ## Networking
-Network is achieved using L3 routing. Because third-party networking plugins (e.g. flannel, calico, etc) don't natively work on Windows Server, existing technology that is built into the Windows and Linux operating systems is relied on. In this L3 networking approach, a /16 subnet is chosen for the cluster nodes, and a /24 subnet is assigned to each worker node. All pods on a given worker node wi ...
+There are several supported network configurations with Windows Server version 1709 and K8s v1.8 including both Layer-3 routed and overlay topologies using third-party network plugins. 
+1. Upstream L3 Routing - IP routes configured in upstream ToR
```
Direct links to each section? :D

bsteciuk replied:
Added direct links in https://github.com/JMesser81/kubernetes.github.io/pull/1

rjmorse commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
-3. RRAS (Routing) Windows feature enabled - Allows routing between NICs on the box, and also "captures" packets that have the destination IP of a POD running on the node. To enable, open "Server Manager". Click on "Roles", "Add Roles". Click "Next". Select "Network Policy and Access Services". Click on "Routing and Remote Access Service" and the underlying checkboxes.
-4. Routes defined pointing to the other pod CIDRs via the "public" NIC - These routes are added to the built-in routing table as shown in **_Route Setup_** section below.
+The following diagram illustrates the Windows Server networking setup for Kubernetes using Upstream L3 Routing Setup:
+![K8s Cluster using L3 Routing with ToR](UpstreamRouting.png)
```
That diagram **really** helps to convey the setup

rjmorse commented:
I didn't see an easy way to find the preview of this page, so here it is: https://deploy-preview-6180--kubernetes-io-master-staging.netlify.com/docs/getting-started-guides/windows/

Maybe this is obvious for those that have updated the docs previously

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+                                         }
+                           }
+                       ]
+}
```
Different sections are definitely needed. This does not apply to all CNI plugins.

alinbalutoiu replied:
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
-2. Run *kube-proxy* executable using the below command
-`.\proxy.exe --v=3 --proxy-mode=userspace --hostname-override=<ip address/hostname of the windows node> --master=<api server location> --bind-address=<ip address of the windows node>`
+## Starting the Linux-based Control Plane
+Use your preferred method to setup and start Kubernetes cluster on Linux or follow the directions given in this [link](https://github.com/Microsoft/SDN/blob/master/Kubernetes/HOWTO-on-prem.md). Please note that Cluster CIDR might need to be updated.
```
Different sections needed here as well.

alinbalutoiu replied:
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

alinbalutoiu commented on docs/getting-started-guides/windows/sample-cniconfig.json:
```diff
@@ -0,0 +1,49 @@
+                                         }
+                           }
+                       ]
+}
```
This file should contain in the filename the network topology for which it applies. The filename suggests that this is valid for all CNI plugins.

JMesser81 replied:
renamed to sample-l2bridge-wincni-config.json

bsteciuk commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+
+**Note:** Windows Server Containers on Kubernetes is an Alpha feature in Kubernetes 1.8.
+
+**Note:** There is one outstanding PR ([51063 Fixes to enable Windows CNI](https://github.com/kubernetes/kubernetes/pull/51063))which has not been merged into v1.8 and is required for Windows CNI to work with kubelet. Users will need to build a private kubelet binary to consume this change. Please refer to these [instructions](https://github.com/Microsoft/SDN/blob/master/Kubernetes/HOWTO-on-prem. ...
```
The note about [51063 Fixes to enable Windows CNI](https://github.com/kubernetes/kubernetes/pull/51063) has been removed in PR to this branch JMesser81/kubernetes.github.io#1

It also makes mention and links to the releases page where pre-built binaries can be found.  It does not address the K8SREPO referencing non-mainline K8S.

bsteciuk commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
 2. Kubenet network plugin setup on the Linux nodes.
-3. Windows Server 2016 (RTM version 10.0.14393 or later).
-4. Docker Version 1.12.2-cs2-ws-beta or later for Windows Server nodes (Linux nodes and Kubernetes control plane can run any Kubernetes supported Docker Version).
+3. Windows Server version 1709 (RTM version 10.0.16299.15 or later).
```
v1.9.0-beta.1 has been released containing the fix for this issue.

bsteciuk commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
 ## Networking
-Network is achieved using L3 routing. Because third-party networking plugins (e.g. flannel, calico, etc) don't natively work on Windows Server, existing technology that is built into the Windows and Linux operating systems is relied on. In this L3 networking approach, a /16 subnet is chosen for the cluster nodes, and a /24 subnet is assigned to each worker node. All pods on a given worker node wi ...
+There are several supported network configurations with Windows Server version 1709 and K8s v1.8 including both Layer-3 routed and overlay topologies using third-party network plugins. 
+1. Upstream L3 Routing - IP routes configured in upstream ToR
```
Added in https://github.com/JMesser81/kubernetes.github.io/pull/1

bsteciuk commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
 The above networking approach is already supported on Linux using a bridge interface, which essentially creates a private network local to the node. Similar to the Windows side, routes to all other pod CIDRs must be created in order to send packets via the "public" NIC.
 
 ### Windows
+Windows supports the CNI network model and uses plugins to interface with the Windows Host Networking Service (HNS) to configure host networking and policy. An administrator creates a local host network using HNS PowerShell commands on each node as documented in the **_Windows Host Setup_** section below.
```
added direct link in https://github.com/JMesser81/kubernetes.github.io/pull/1

bsteciuk commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+Windows supports the CNI network model and uses plugins to interface with the Windows Host Networking Service (HNS) to configure host networking and policy. An administrator creates a local host network using HNS PowerShell commands on each node as documented in the **_Windows Host Setup_** section below.
+
+#### Upstream L3 Routing Topology
+In this topology, networking is achieved using L3 routing with static IP routes configured in an upstream Top of Rack (ToR) switch/router. Each cluster node is connected to the management network with a host IP. Additionally, each node uses a local 'l2bridge' network with a Pod CIDR assigned. All pods on a given worker node will be connected to the POD CIDR subnet ('l2bridge' network). In order t ...
```
fixed pod/POD/Pod nit in https://github.com/JMesser81/kubernetes.github.io/pull/1

bsteciuk commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
 
 ## Setting up Windows Server Containers on Kubernetes
-To run Windows Server Containers on Kubernetes, you'll need to set up both your host machines and the Kubernetes node components for Windows and setup Routes for Pod communication on different nodes.
+To run Windows Server Containers on Kubernetes, you'll need to set up both your host machines and the Kubernetes node components for Windows and depending on your network topology, setup Routes for Pod communication on different nodes.
```
addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

bsteciuk commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
-2. Build *kube-proxy*
-   1. Linux/MacOS: `KUBE_BUILD_PLATFORMS=windows/amd64 make WHAT=cmd/kube-proxy`
-   2. Windows: `go build cmd/kube-proxy/proxy.go`
+2. Configure Linux Master node using steps [here](https://github.com/Microsoft/SDN/blob/master/Kubernetes/HOWTO-on-prem.md)
```
Added direct links in https://github.com/JMesser81/kubernetes.github.io/pull/1

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
 2. Kubenet network plugin setup on the Linux nodes.
-3. Windows Server 2016 (RTM version 10.0.14393 or later).
-4. Docker Version 1.12.2-cs2-ws-beta or later for Windows Server nodes (Linux nodes and Kubernetes control plane can run any Kubernetes supported Docker Version).
+3. Windows Server version 1709 (RTM version 10.0.16299.15 or later).
```
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+There are several supported network configurations with Windows Server version 1709 and K8s v1.8 including both Layer-3 routed and overlay topologies using third-party network plugins. 
+1. Upstream L3 Routing - IP routes configured in upstream ToR
+2. Host-Gateway - IP routes configured on each host
+3. OVN & OVS with Overlay - OVS switch extension and OVN controller creates VXLAN overlay network
```
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+4. [Future] Overlay - VXLAN or IP-in-IP encapsulation using Flannel
+5. [Future] Layer-3 Routing with BGP (Calico)
+
+## CNI Plugins
```
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
+#### Host-Gateway Topology
+This topology is similar to the Upstream L3 Routing topology with the only difference being that static IP routes are configured directly on each cluster node and not in the upstream ToR. Each node uses a local 'l2bridge' network with a Pod CIDR assigned as before and has routing table entries for all other Pod CIDR subnets assigned to the remote cluster nodes.
+
+#### Overlay using OVN controller and OVS Switch Extension
```
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
-To run Windows Server Containers on Kubernetes, you'll need to set up both your host machines and the Kubernetes node components for Windows and setup Routes for Pod communication on different nodes.
+To run Windows Server Containers on Kubernetes, you'll need to set up both your host machines and the Kubernetes node components for Windows and depending on your network topology, setup Routes for Pod communication on different nodes.
 
 ### Host Setup
```
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
-2. Build *kube-proxy*
-   1. Linux/MacOS: `KUBE_BUILD_PLATFORMS=windows/amd64 make WHAT=cmd/kube-proxy`
-   2. Windows: `go build cmd/kube-proxy/proxy.go`
+2. Configure Linux Master node using steps [here](https://github.com/Microsoft/SDN/blob/master/Kubernetes/HOWTO-on-prem.md)
```
Addressed in https://github.com/JMesser81/kubernetes.github.io/pull/1

idvoretskyi commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,139 +1,144 @@
-**Note:** Windows Server Containers on Kubernetes is an Alpha feature in Kubernetes 1.5.
+ The Kubernetes control plane (API Server, Scheduler, Controller Manager, etc) continue to run on Linux, while the kubelet and kube-proxy can be run on Windows Server version 1709.
+
+**Note:** Windows Server Containers on Kubernetes is an Alpha feature in Kubernetes 1.8.
```
It should be beta in Kubernetes 1.9 - please, update.

michmike replied:
@idvoretskyi  we have a pull on this ticket with the updated docs. please reference https://github.com/JMesser81/kubernetes.github.io/pull/1

idvoretskyi commented:
@JMesser81 since we're targeting this feature to be delivered as beta to Kubernetes 1.9, please, update all the Kubernetes version references to 1.9 (there are several mentions of v1.5 and v1.8).

cc @zacharysarah @kubernetes/sig-windows-feature-requests

bsteciuk commented on docs/getting-started-guides/windows/index.md:
````diff
@@ -1,144 +1,256 @@
+```
+K8SREPO="github.com/kubernetes/kubernetes"
+go get -d $K8SREPO
+# Note: the above command may spit out a message about 
````
The comments in this code block look to be causing a formatting problem (comments showing up as headers) on the docs preview site https://deploy-preview-6180--kubernetes-io-vnext-staging.netlify.com/docs/getting-started-guides/windows/

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
````diff
@@ -17,28 +17,7 @@ The Kubernetes control plane (API Server, Scheduler, Controller Manager, etc) co
-
-# You will find the output binaries under the folder _output/local/bin/windows/
-```
+If you wish to build the code yourself, please follow the [Compiling Kubernetes Binaries instructions](https://github.com/microsoft/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/compiling-kubernetes-binaries.md)
````
Any reason for this? I thought we discussed it here: https://github.com/JMesser81/kubernetes.github.io/pull/1#discussion_r154569768
I agree with Michael on this. I also prefer to have the steps on how to build K8s for windows in this page.

JMesser81 replied:
Sure. I'll place the build instructions back on this main index page

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -99,21 +78,21 @@ To run Windows Server Containers on Kubernetes, you'll need to set up both your
 
 1. Linux hosts should be setup according to their respective distro documentation and the requirements of the Kubernetes version you will be using. 
-2. Configure Linux Master node using steps [here](https://github.com/Microsoft/SDN/blob/master/Kubernetes/HOWTO-on-prem.md#prepare-the-master)
+2. Configure Linux Master node using steps [here](https://github.com/gkudra-msft/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/creating-a-linux-master.md)
```
Don't put references to fork. You should change this with the official repository even if it's not merged yet in that, you know where the file would be after it will be merged.

JMesser81 replied:
I thought I fixed previously... will address. thanks

bsteciuk commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,144 +1,235 @@
+##### Windows Host Setup
+
+
+1. Windows Server container host running the required Windows Server and Docker versions. Follow the setup instructions outlined by this help topic: https://docs.microsoft.com/en-us/virtualization/windowscontainers/quick-start/quick-start-windows-server.
```
this should probably be a link instead of just text

JMesser81 commented:
I rebased to before I merged with upstream/master. I then merged changes from release-1.9 into my fork / branch. It seems to have picked up a few extra files but the history shows these are included in between when PR#2 was merged to PR#1. Any idea how to clean this up a bit more. I think I've addressed all other issues.

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -55,19 +56,19 @@ There are several supported network configurations with Kubernetes v1.9 on Windo
 1. [Upstream L3 Routing](#upstream-l3-routing-topology) - IP routes configured in upstream ToR
 2. [Host-Gateway](#host-gateway-topology) - IP routes configured on each host
 3. [Open vSwitch (OVS) & Open Virtual Network (OVN) with Overlay](#overlay-using-ovn-controller-and-ovs-switch-extension-topology) - overlay networks (supports STT and Geneve tunneling types)
```
Can you please also replace 
```
#overlay-using-ovn-controller-and-ovs-switch-extension-topology
```
with
```
#using-ovn-with-ovs
```
Thanks!

JMesser81 replied:
yup. just did.

alinbalutoiu commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -58,7 +58,7 @@ There are several supported network configurations with Kubernetes v1.9 on Windo
 1. [Upstream L3 Routing](#upstream-l3-routing-topology) - IP routes configured in upstream ToR
 2. [Host-Gateway](#host-gateway-topology) - IP routes configured on each host
-3. [Open vSwitch (OVS) & Open Virtual Network (OVN) with Overlay](#overlay-using-ovn-controller-and-ovs-switch-extension-topology) - overlay networks (supports STT and Geneve tunneling types)
+3. [Open vSwitch (OVS) & Open Virtual Network (OVN) with Overlay](#Using-OVN-with-OVS) - overlay networks (supports STT and Geneve tunneling types)
```
Can you please replace 
```
#Using-OVN-with-OVS
```
with
```
#using-ovn-with-ovs
```
It's case sensitive and should be with lowercase letters. Thanks!

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,144 +1,262 @@
 2. Kubenet network plugin setup on the Linux nodes.
-3. Windows Server 2016 (RTM version 10.0.14393 or later).
-4. Docker Version 1.12.2-cs2-ws-beta or later for Windows Server nodes (Linux nodes and Kubernetes control plane can run any Kubernetes supported Docker Version).
+3. Windows Server 2016 RTM or later. Windows Server version 1709 or later is preferred; unlocks key capabilities like shared network namespace.
```
Minor nit: "is preferred; _**it**_ unlocks key...". Semicolons should always be replaceable by periods.

JMesser81 replied:
fixed

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -440,5 +440,5 @@ Some of these limitations will be addressed by the community in future releases
 - Hyper-V Containers are not supported
 
-[!Warning]   
+![Warning]   
```
This is a MSDN syntax; remove this line and either make the content a bullet point like the above issues or a block-quote by prefixing with `> `.

JMesser81 replied:
updated

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,144 +1,262 @@
+More detailed instructions can be found [here](https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/getting-started-kubernetes-windows.md)
 
-To build the *kubelet*, run:
+**Windows CNI Config Example**
```
Add two spaces at the end of a line to make it an actual new line.

JMesser81 replied:
done

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,144 +1,262 @@
-* make (if using Linux or MacOS)
-* Important notes and other dependencies are listed [here](https://git.k8s.io/community/contributors/devel/development.md#building-kubernetes-on-a-local-osshell-environment)
+1. Windows Server container host running the required Windows Server and Docker versions. Follow the setup instructions outlined by this help topic: https://docs.microsoft.com/en-us/virtualization/windowscontainers/quick-start/quick-start-windows-server.
+2. Build or download kubelet.exe, kube-proxy.exe, and kubectl.exe using instructions found [here](https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/compiling-kubernetes-binaries.md)
```
Should this be a same-document link to the build instructions?

JMesser81 replied:
Yes. Updated

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
````diff
@@ -1,144 +1,262 @@
+
+2. Run the following commands to build kubelet and kube-proxy:
+
+```code
````
change it to ` ```bash ` to look prettier

JMesser81 replied:
fixed

gkudra-msft commented on docs/getting-started-guides/windows/index.md:
````diff
@@ -1,144 +1,262 @@
+sudo apt-get install curl git build-essential docker.io conntrack
+```
+
+2. Run the following commands to build kubelet and kube-proxy:
````
Looks like it's resetting the numbering because of the code-block in step 1. Try indenting it all (including the triple-backticks) two spaces? Not sure.

JMesser81 replied:
addressed

alinbalutoiu commented:
I believe you still have a few commits which don't belong to this PR, such as:
https://github.com/kubernetes/website/pull/6180/commits/c05678752d3b481e2907bc53d3971bb49eab6609
https://github.com/kubernetes/website/pull/6180/commits/b84ac59624b625e6534ccd97bb4ba65e51b441e4
https://github.com/kubernetes/website/pull/6180/commits/31e4dbdc25a60e4584ce01a6b1915e13ac63bc67
and a few others. Can you please remove those commits from this PR? Thanks!

JMesser81 commented:
Addressed all feedback and removed non-related commits (merged in by mistake previously)

alinbalutoiu commented:
Hi Jason, can you please also remove unrelated commits such as:
https://github.com/kubernetes/website/pull/6180/commits/5d39cfeae41b3237a5e1247bc1c1f98e0727c5fd
https://github.com/kubernetes/website/pull/6180/commits/47eed4346e4491c9a63c2e0cb76bdd37bff5677c
And a few others, I don't think they belong to this PR.

michmike commented:
@JMesser81 i think the formatting is off now.  something happened after `More detailed build instructions will be maintained and kept up to date [here](https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/compiling-kubernetes-binaries.md)
`

JMesser81 commented:
@michmike  - i'm not seeing any weird formatting when I render here: https://deploy-preview-6180--kubernetes-io-vnext-staging.netlify.com/docs/getting-started-guides/windows/ That section looks fine to me, can you be more specific?

michmike commented:
@JMesser81 the deploy preview does look perfect. i was looking at this view which in the past (up to yesterday) was accurate
https://github.com/JMesser81/kubernetes.github.io/blob/61bec6ca2dd7c0f90b1aaa9d6773bd1e2b570e74/docs/getting-started-guides/windows/index.md

michmike commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r1, 6 of 3034 files at r2, 2 of 3 files at r3, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 32 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/website/6180#-:-L-byjw-0pPy-_mRR7Cj:bg7igk9)*
<!-- Sent from Reviewable.io -->

zacharysarah commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,144 +1,261 @@
-title: Windows Server Containers
+title: Using Windows Server Containers in Kubernetes
 ---
+**Note:** These instructions were recently updated based on Windows Server platform enhancements and the Kubernetes v1.9 release
```
```
**Note:** These instructions were recently updated based on Windows Server platform enhancements and the Kubernetes v1.9 release.
{: .note}
```
So:
> **Note:** These instructions were recently updated based on Windows Server platform enhancements and the Kubernetes v1.9 release.
> {: .note}

zacharysarah commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,144 +1,261 @@
-**Note:** Windows Server Containers on Kubernetes is an Alpha feature in Kubernetes 1.5.
+The Kubernetes control plane (API Server, Scheduler, Controller Manager, etc) continue to run on Linux, while the kubelet and kube-proxy can be run on Windows Server 2016 or later
+
+**Note:** Windows Server Containers on Kubernetes is a Beta feature in Kubernetes v1.9
```
```
**Note:** Windows Server Containers on Kubernetes is a Beta feature in Kubernetes v1.9.
{: .note}
```
So:
> **Note:** Windows Server Containers on Kubernetes is a Beta feature in Kubernetes v1.9.
> {: .note}

zacharysarah commented on docs/getting-started-guides/windows/index.md:
````diff
@@ -1,144 +1,261 @@
+    # You will find the output binaries under the folder _output/local/bin/windows/
+```
+
+More detailed build instructions will be maintained and kept up to date [here](https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/compiling-kubernetes-binaries.md)
````
Add a period at the end.
```
More detailed build instructions will be maintained and kept up to date [here](https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/compiling-kubernetes-binaries.md).
```

zacharysarah commented on docs/getting-started-guides/windows/index.md:
```diff
@@ -1,144 +1,261 @@
+6. [Only required for #2 Host-Gateway mode] Add static routes on Windows host using this script [AddRoutes.ps1](https://github.com/Microsoft/SDN/blob/master/Kubernetes/windows/AddRoutes.ps1)
 
-**kubelet**
+More detailed instructions can be found [here](https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/getting-started-kubernetes-windows.md)
```
Add a period at the end.
```
More detailed instructions can be found [here](https://github.com/MicrosoftDocs/Virtualization-Documentation/blob/live/virtualization/windowscontainers/kubernetes/getting-started-kubernetes-windows.md).
```

michmike commented:
thanks for the review @zacharysarah !

zacharysarah commented:
@JMesser81 👋 If you'll un-check and re-check the box to allow edits from maintainers (it's a workaround for a broken GitHub UI component), I'm happy to make edits for you. Otherwise, please go ahead and make the changes yourself. 👍 
![image](https://user-images.githubusercontent.com/3210446/33684733-e32cdb2a-da94-11e7-8baa-7b1620d4a288.png)

JMesser81 commented:
Sorry for the delay. Addressed comments.
